lib/dataset/bin.cpp

```diff
@@ -30,74 +31,186 @@ namespace scipp::dataset {
+                         std::nullopt) const override {
+    // Up until here the output was viewed with same bin index ranges as input.
+    // Now setup the desired final bin indices.
+    const auto dims = dims_override.value_or(m_dims);
+    auto output_dims = merge(m_output_bin_sizes.dims(), dims);
+    return makeVariable<scipp::index>(
+        output_dims, units::none,
+        Values(flatten_subbin_sizes(m_output_bin_sizes, dims.volume())));
+  }
+
+  Dimensions m_dims;
+  Variable m_indices;
+
+private:
+  Variable m_output_bin_sizes;
+  Variable m_offsets;
+  Variable m_filtered_input_bin_ranges;
+  scipp::index m_total_size;
+};
+
+class TwoStageMapper : public Mapper {
+public:
+  TwoStageMapper(SingleStageMapper &&stage1_mapper,
+                 SingleStageMapper &&stage2_mapper)
+      : m_stage1_mapper(std::move(stage1_mapper)),
+        m_stage2_mapper(std::move(stage2_mapper)) {}
+
+  Variable _do_bin(const Variable &var) const override {
+    // Note how by having the virtual call on the Variable level we avoid
+    // making the temporary buffer for the whole content buffer (typically a
+    // DataArray), but instead just for one of the content buffer's columns
+    // at a time.
+    Variable content = m_stage1_mapper._do_bin(var);
+    Variable indices = m_stage2_mapper.m_indices.bin_indices();
+    return m_stage2_mapper._do_bin(
+        make_bins_no_validate(indices, content.dims().inner(), content));
+  }
+
+  Variable bin_sizes(const std::optional<Dimensions> &dims_override =
+                         std::nullopt) const override {
+    const auto dims = dims_override.value_or(m_stage1_mapper.m_dims);
+    return fold(
+        flatten(m_stage2_mapper.bin_sizes(),
+                std::vector<Dim>{Dim::InternalBinCoarse, Dim::InternalBinFine},
+                Dim::InternalSubbin)
+            .slice({Dim::InternalSubbin, 0, dims.volume()}),
+        Dim::InternalSubbin, dims);
+  }
+
+private:
+  SingleStageMapper m_stage1_mapper;
+  SingleStageMapper m_stage2_mapper;
+};
+
+template <class Builder>
+std::unique_ptr<Mapper> make_mapper(const Variable &indices,
+                                    const Builder &builder) {
+  const auto dims = builder.dims();
+  if (use_two_stage_remap(builder)) {
+    // There are many output bins. Mapping directly would lead to excessive
+    // number of cache misses as well as potential false-sharing problems
+    // between threads. We therefore map in two stages. This requires an
+    // additional temporary buffer with size given by the number of events,
+    // but has proven to be faster in practice.
+    scipp::index chunk_size =
+        floor(sqrt(builder.nbin().template value<scipp::index>()));
+    const auto chunk = astype(scipp::index{chunk_size} * units::none,
+                              indices.bin_buffer<Variable>().dtype());
+    Variable fine_indices = indices;
+    auto indices_ = floor_divide(indices, chunk);
+    fine_indices %= chunk;
```

Comment:
```suggestion
    auto fine_indices = indices % chunk;
    const auto coarse_indices = floor_divide(indices, chunk);
```
and rename usages of `indices_`.

Reply:
How so? You already copy the indices.

Reply:
That is is shallow copy, it should be sharing the buffer?

Reply:
But then it modifies the input indices even though they are declared `const`. Is this a problem?

Reply:
Not in this case, since we know that they are not used any more.

Reply:
Then I'd feel safer if you changed the function to accept an rvalue so it's clear on the calling site that the indices cannot be used afterwards.